pkg/autodiscovery/listeners/cloudfoundry.go

```diff
@@ -0,0 +1,235 @@
+			//       which changes by redeploying
+			delete(notSeen, strId)
+			continue
+		}
+		// TODO: handle container state?
+		svc := l.createService(id, firstRun)
+		l.newService <- svc
+	}
+
+	for adId := range notSeen {
+		l.m.RLock()
+		l.delService <- l.services[adId]
+		l.m.RUnlock()
+		l.m.Lock()
+		delete(l.services, adId)
+		l.m.Unlock()
+	}
+}
+
+func (l *CloudFoundryListener) createService(adId cloudfoundry.ADIdentifier, firstRun bool) *CloudFoundryService {
+	l.m.Lock()
+	defer l.m.Unlock()
+	var crTime integration.CreationTime
+	if firstRun {
+		crTime = integration.Before
+	} else {
+		crTime = integration.After
+	}
+
+	var svc *CloudFoundryService
+	aLRP := adId.GetActualLRP()
+	if aLRP == nil {
+		// non-container service
+		svc = &CloudFoundryService{
+			adIdentifier:   adId,
+			containerIPs:   map[string]string{}, // TODO
+			containerPorts: []ContainerPort{},   // TODO
+			creationTime:   crTime,
+		}
+	} else {
+		// container service => we need one service per container instance
+		ips := map[string]string{CF_SERVICE_CONTAINER_IP: aLRP.ContainerIP}
+		ports := []ContainerPort{}
+		for _, p := range aLRP.Ports {
+			ports = append(ports, ContainerPort{
+				// TODO: because of how configresolver.getPort works, we can't use e.g. port_8080, so we use port_p8080
+				// can we change that logic?
+				Name: fmt.Sprintf("p%d", p),
+				Port: int(p),
+			})
+		}
+		svc = &CloudFoundryService{
+			adIdentifier:   adId,
+			containerIPs:   ips,
+			containerPorts: ports,
+			creationTime:   crTime,
+		}
+	}
+	l.services[adId.String()] = svc
+	return svc
+}
+
+func (l *CloudFoundryListener) getAllADIdentifiers() []cloudfoundry.ADIdentifier {
+	ret := []cloudfoundry.ADIdentifier{}
+	for _, dLRP := range l.bbsCache.GetDesiredLRPs() {
+		for adName, _ := range dLRP.EnvAD {
+			if _, ok := dLRP.EnvVcapServices[adName]; ok {
+				// if it's in VCAP_SERVICES, it's a non-container service and we want one instance per App
+				ret = append(ret, cloudfoundry.NewADNonContainerIdentifier(dLRP, adName))
+			} else {
+				// if it's not in VCAP_SERVICES, it's a container service and we want one instance per container
+				for _, aLRP := range l.bbsCache.GetActualLRPsFor(dLRP.AppGUID) {
+					ret = append(ret, cloudfoundry.NewADContainerIdentifier(dLRP, adName, aLRP))
+				}
+			}
+		}
+	}
+	return ret
+}
+
+// Stop queues a shutdown of CloudFoundryListener
+func (l *CloudFoundryListener) Stop() {
+	l.stop <- true
+}
+
+// GetEntity returns the unique entity name linked to that service
+func (s *CloudFoundryService) GetEntity() string {
+	return s.adIdentifier.String()
+}
+
+// GetTaggerEntity returns the unique entity name linked to that service
+func (s *CloudFoundryService) GetTaggerEntity() string {
+	return s.adIdentifier.String()
+}
+
+// GetADIdentifiers returns a set of AD identifiers for a container.
+func (s *CloudFoundryService) GetADIdentifiers() ([]string, error) {
+	return []string{s.adIdentifier.String()}, nil
```

Comment:
This must match the service identifiers used in the tagger - if we're going to add cloud foundry tagger collector.

Reply:
I think we will add that eventually; matching the identifiers in the tagger shouldn't be a problem, IMO.

Reply:
Is it supposed to be the same ID I'm building here then ? https://github.com/DataDog/datadog-agent/pull/4921/files#diff-4fcf38f3713ce8d1e3c3b3d7542f8771R73